zkevm-circuits/src/evm_circuit/execution/call.rs

```diff
@@ -0,0 +1,718 @@
+            region,
+            offset,
+            F::from(gas.low_u64()),
+            F::from(gas_available - gas_available / 64),
+        )?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{
+        test::{run_test_circuit_complete_fixed_table, run_test_circuit_incomplete_fixed_table},
+        witness::block_convert,
+    };
+    use eth_types::bytecode;
+    use eth_types::{bytecode::Bytecode, evm_types::OpcodeId, geth_types::Account};
+    use eth_types::{Address, ToWord, Transaction, Word};
+    use itertools::Itertools;
+    use std::default::Default;
+
+    #[derive(Clone, Copy, Debug, Default)]
+    struct Stack {
+        gas: u64,
+        value: Word,
+        cd_offset: u64,
+        cd_length: u64,
+        rd_offset: u64,
+        rd_length: u64,
+    }
+
+    fn caller(stack: Stack, caller_is_success: bool) -> Account {
+        let terminator = if caller_is_success {
+            OpcodeId::RETURN
+        } else {
+            OpcodeId::REVERT
+        };
+
+        // Call twice for testing both cold and warm access
+        let bytecode = bytecode! {
+            PUSH32(Word::from(stack.rd_length))
+            PUSH32(Word::from(stack.rd_offset))
+            PUSH32(Word::from(stack.cd_length))
+            PUSH32(Word::from(stack.cd_offset))
+            PUSH32(stack.value)
+            PUSH32(Address::repeat_byte(0xff).to_word())
+            PUSH32(Word::from(stack.gas))
+            CALL
+            PUSH32(Word::from(stack.rd_length))
+            PUSH32(Word::from(stack.rd_offset))
+            PUSH32(Word::from(stack.cd_length))
+            PUSH32(Word::from(stack.cd_offset))
+            PUSH32(stack.value)
+            PUSH32(Address::repeat_byte(0xff).to_word())
+            PUSH32(Word::from(stack.gas))
+            CALL
+            PUSH1(0)
+            PUSH1(0)
+            .write_op(terminator)
+        };
+
+        Account {
+            address: Address::repeat_byte(0xfe),
+            balance: Word::from(10).pow(20.into()),
+            code: bytecode.to_vec().into(),
+            ..Default::default()
+        }
+    }
+
+    fn callee(code: Bytecode) -> Account {
+        let code = code.to_vec();
+        let is_empty = code.is_empty();
+        Account {
+            address: Address::repeat_byte(0xff),
+            code: code.into(),
+            nonce: if is_empty { 0 } else { 1 }.into(),
+            balance: if is_empty { 0 } else { 0xdeadbeefu64 }.into(),
+            ..Default::default()
+        }
+    }
+
+    fn test_ok(caller: Account, callee: Account, use_complete_fixed_table: bool) {
+        let tx = Transaction {
+            to: Some(caller.address),
+            gas: 100000.into(),
+            transaction_index: Some(0.into()),
+            ..Default::default()
+        };
+        let block_data = bus_mapping::mock::BlockData::new_from_geth_data(
+            mock::new(vec![caller, callee], vec![tx]).unwrap(),
+        );
```

Comment:
Can we try to move this to the new `TestContext` struct for the setup?? 
I think might help to discover possible missing things!